src/storage/src/render/upsert.rs

```diff
@@ -146,18 +147,74 @@ where
+        // ## Correctness
+        //
+        // The method is correct because a well formed upsert snapshot will have for each key:
+        // - Zero or one updates of the form (cur_value, +1)
+        // - Zero or more pairs of updates of the form (prev_value, +1), (prev_value, -1)
+        //
+        // We are interested in extracting the cur_value of each key and discard all prev_values
+        // that might be included in the stream. Since the history of prev_values always comes in
+        // pairs, computing the XOR of those is always going to cancel their effects out. Also,
+        // since XOR is commutative this property is true independent of the order. The same is
+        // true for the summations of the length and checksum since the sum will contain the
+        // unrelated values zero times.
+        //
+        // Therefore the accumulators will end up precisely in one of two states:
+        // 1. diff == 0, checksum == 0, value == [0..] => the key is not present
+        // 2. diff == 1, checksum == checksum(cur_value) value == cur_value => the key is present
+        //
+        // ## Robustness
+        //
+        // In the absense of bugs, accumulating the diff and checksum is not required since we know
+        // that a well formed snapshot always satisfies XOR(bincode(values)) == bincode(cur_value).
+        // However bugs may happen and so storing 16 more bytes per key to have a very high
+        // guarantee that we're not decoding garbage is more than worth it.
+        #[allow(clippy::disallowed_types)]
+        let mut snapshot = std::collections::HashMap::<
+            UpsertKey,
+            (Vec<u8>, Wrapping<i64>, Wrapping<i64>, Wrapping<i64>),
+        >::new();
+        let mut buf = vec![];
+        let enc_opts = bincode::DefaultOptions::new().allow_trailing_bytes();
         while let Some(event) = previous.next_mut().await {
             match event {
                 AsyncEvent::Data(_cap, data) => {
-                    snapshot.extend(
-                        data.drain(..)
-                            .filter(|(_row, ts, _diff)| !resume_upper.less_equal(ts))
-                            .map(|(row, _ts, diff)| (row, diff)),
-                    );
+                    for ((key, value), ts, diff) in data.drain(..) {
+                        #[allow(clippy::as_conversions)]
+                        if !resume_upper.less_equal(&ts) {
+                            buf.clear();
+                            enc_opts.serialize_into(&mut buf, &value).unwrap();
+                            let len = i64::try_from(buf.len()).unwrap();
+
+                            let entry = snapshot.entry(key).or_default();
+                            let (value_xor, len_sum, checksum_sum, diff_sum) = entry;
+
+                            *diff_sum += diff;
+                            *len_sum += len.wrapping_mul(diff);
+                            *checksum_sum += (seahash::hash(&buf) as i64).wrapping_mul(diff);
```

Comment:
What happens in case there is a hash collision?

Reply:
Nothing in particular, it is fine for two values to have the same hash. In order for this computation to produce an incorrect snapshot many stars would have to align. We'd have to have a bug that leaves the shard with an insertion of a value `v1` and a retraction of a value `v2` (without their corresponding pairs) and to happen that `checksum(v1) == checksum(v2)` and `len(v1) == len(v2)`.